xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -318,81 +506,45 @@ private LdsUpdate processServerSideListener(Listener listener)
             "HttpFilter [" + filterName + "] is not optional and has no typed config");
       }
     }
-    return parseRawFilterConfig(filterName, httpFilter.getTypedConfig(), isOptional, false);
-  }
-
-  @Nullable // Returns null if the filter should be ignored.
-  private static StructOrError<FilterConfig> parseRawFilterConfig(
-      String filterName, Any anyConfig, Boolean isOptional, boolean isOverrideConfig) {
-    checkArgument(
-        isOptional != null || isOverrideConfig, "isOptional can't be null for top-level config");
-    String typeUrl = anyConfig.getTypeUrl();
-    if (isOverrideConfig) {
-      isOptional = false;
-      if (typeUrl.equals(TYPE_URL_FILTER_CONFIG)) {
-        io.envoyproxy.envoy.config.route.v3.FilterConfig filterConfig;
-        try {
-          filterConfig =
-              anyConfig.unpack(io.envoyproxy.envoy.config.route.v3.FilterConfig.class);
-        } catch (InvalidProtocolBufferException e) {
-          return StructOrError.fromError(
-              "HttpFilter [" + filterName + "] contains invalid proto: " + e);
-        }
-        isOptional = filterConfig.getIsOptional();
-        anyConfig = filterConfig.getConfig();
-        typeUrl = anyConfig.getTypeUrl();
-      }
-    }
-    Message rawConfig = anyConfig;
-    if (anyConfig.getTypeUrl().equals(TYPE_URL_TYPED_STRUCT)) {
+    Message rawConfig = httpFilter.getTypedConfig();
+    String typeUrl = httpFilter.getTypedConfig().getTypeUrl();
+    if (typeUrl.equals(TYPE_URL_TYPED_STRUCT)) {
       TypedStruct typedStruct;
       try {
-        typedStruct = anyConfig.unpack(TypedStruct.class);
+        typedStruct = httpFilter.getTypedConfig().unpack(TypedStruct.class);
       } catch (InvalidProtocolBufferException e) {
         return StructOrError.fromError(
             "HttpFilter [" + filterName + "] contains invalid proto: " + e);
       }
       typeUrl = typedStruct.getTypeUrl();
       rawConfig = typedStruct.getValue();
     }
-    Filter filter = FilterRegistry.getDefaultRegistry().get(typeUrl);
-    if (filter == null) {
+    Filter filter = filterRegistry.get(typeUrl);
+    if (filter == null || (isForClient && !(filter instanceof ClientInterceptorBuilder))
+        || (!isForClient && !(filter instanceof ServerInterceptorBuilder))) {
       if (isOptional) {
         return null;
       } else {
         return StructOrError.fromError(
-            "HttpFilter [" + filterName + "] is not optional and has an unsupported config type: "
-                + typeUrl);
+            "HttpFilter [" + filterName + "](" + typeUrl + ") is required but unsupported for "
```

Comment:
I noticed you don't have a test to verify this error generation. So the original bug on lines 523-524 went undetected.

Reply:
Actually it is tested. See tests started with `parseHttpFilter_`:

- The RouterFilter is supported for both client and server, so we use it to cover the branch `isForClient && (filter instanceof ClientInterceptorBuilder)` and `!isForClient && (filter instanceof ServerInterceptorBuilder)`.
- The FaultFilter is supported for client only, so we used it to cover the branch `!isForClient && !(filter instanceof ServerInterceptorBuilder)`
- The RbacFilter is supported for server only, so we used it to cover the branch `isForClient && !(filter instanceof ClientInterceptorBuilder)`

The four branches covered all forClient + supported/unsupported and forServer + supported/unsupported cases.

Reply:
Eliminated the boolean zen in line 523-524, `filter == null ||` is not needed, its included in `instanceof`.